### PR TITLE
[AArch64] Implements FP8 SVE intrinsics for dot-product

### PR DESCRIPTION
**Pull request by momchil-velikov**
This patch adds the following intrinsics:

* 8-bit floating-point dot product to single-precision.

  // Only if (__ARM_FEATURE_SVE2  && __ARM_FEATURE_FP8DOT4) || __ARM_FEATURE_SSVE_FP8DOT4
  svfloat32_t svdot[_f32_mf8]_fpm(svfloat32_t zda, svmfloat8_t zn, svmfloat8_t zm, fpm_t fpm);
  svfloat32_t svdot[_n_f32_mf8]_fpm(svfloat32_t zda, svmfloat8_t zn, mfloat8_t zm, fpm_t fpm);

* 8-bit floating-point indexed dot product to single-precision.

  // Only if (__ARM_FEATURE_SVE2  && __ARM_FEATURE_FP8DOT4) || __ARM_FEATURE_SSVE_FP8DOT4
  svfloat32_t svdot_lane[_f32_mf8]_fpm(svfloat32_t zda, svmfloat8_t zn, svmfloat8_t zm,
                                       uint64_t imm0_3, fpm_t fpm);

* 8-bit floating-point dot product to half-precision.

  // Only if (__ARM_FEATURE_SVE2  && __ARM_FEATURE_FP8DOT2) || __ARM_FEATURE_SSVE_FP8DOT2
  svfloat16_t svdot[_f16_mf8]_fpm(svfloat16_t zda, svmfloat8_t zn, svmfloat8_t zm, fpm_t fpm);
  svfloat16_t svdot[_n_f16_mf8]_fpm(svfloat16_t zda, svmfloat8_t zn, mfloat8_t zm, fpm_t fpm);

* 8-bit floating-point indexed dot product to half-precision.

  // Only if (__ARM_FEATURE_SVE2  && __ARM_FEATURE_FP8DOT2) || __ARM_FEATURE_SSVE_FP8DOT2
  svfloat16_t svdot_lane[_f16_mf8]_fpm(svfloat16_t zda, svmfloat8_t zn, svmfloat8_t zm,
                                       uint64_t imm0_7, fpm_t fpm);

### CODE REVIEW

**Review by jthackray — APPROVED** [review]
LGTM

File: llvm/lib/Target/AArch64/SVEInstrFormats.td [hunk 1]
```diff
@@ -10877,24 +10923,37 @@ class sve_fp8_dot_indexed<bits<4> opc, ZPRRegOp dst_ty, Operand iop_ty, string m
...
-multiclass sve2_fp8_dot_indexed_h<string asm>{
-  def NAME : sve_fp8_dot_indexed<{0, ?, ?, ?}, ZPR16, VectorIndexH, asm> {
+multiclass sve2_fp8_dot_indexed_h<string asm, SDPatternOperator op> {
+  def NAME : sve_fp8_dot_indexed<{0, ?, ?, ?}, ZPR16, VectorIndexH32b, asm> {
     bits<3> iop;
 
     let Inst{20-19} = iop{2-1};
     let Inst{11}    = iop{0};
+
+    let mayLoad  = 1;
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/SVEInstrFormats.td:10939]: Do we need to repeat this here if the class sve_fp8_dot_indexed already have:
 let mayLoad  = 1;
 let mayStore = 0;

File: clang/include/clang/Basic/arm_sve_sme_incl.td [hunk 2]
```diff
@@ -105,6 +105,7 @@ include "arm_immcheck_incl.td"
 // N: svfloat64_t
 // $: svbfloat16_t
 // ~: svmfloat8_t
+// !: mfloat8_t (splat to svmfloat8_t)
```
[the remark this answers is not in the document]
Comment (momchil-velikov) [reply clang/include/clang/Basic/arm_sve_sme_incl.td]: Done

File: clang/include/clang/Basic/arm_sve.td [hunk 3]
```diff
@@ -2447,3 +2447,39 @@ let SVETargetGuard = "sve2,faminmax", SMETargetGuard = "sme2,faminmax" in {
...
+  def SVFDOT_N_2WAY :  SInst<"svdot[_n_f16_mf8]_fpm", "dd~!>", "h", MergeNone, "aarch64_sve_fp8_fdot", [VerifyRuntimeMode, SetsFPMR]>;
+
+  // 8-bit floating-point dot product to half-precision (indexed)
+  def SVFDOT_LANE_2WAY :  SInst<"svdot_lane[_f16_mf8]_fpm", "dd~~i>", "h", MergeNone, "aarch64_sve_fp8_fdot_lane", [VerifyRuntimeMode, SetsFPMR], [ImmCheck<3, ImmCheck0_7>]>;
+}
+
+let SVETargetGuard = "sve2,fp8dot4", SMETargetGuard ="sme,ssve-fp8dot4" in {
+  // 8-bit floating-point dot product to single-precision (vectors)
+  def SVFDOT_4WAY   : SInst<"svdot[_f32_mf8]_fpm",   "dd~~>", "f", MergeNone, "aarch64_sve_fp8_fdot", [VerifyRuntimeMode, SetsFPMR]>;
+  def SVFDOT_N_4WAY : SInst<"svdot[_n_f32_mf8]_fpm", "dd~!>", "f", MergeNone, "aarch64_sve_fp8_fdot", [VerifyRuntimeMode, SetsFPMR]>;
```
[the remark this answers is not in the document]
Comment (momchil-velikov) [reply clang/include/clang/Basic/arm_sve.td:2486]: They were accidentally omitted, I'm adding them via https://github.com/ARM-software/acle/pull/369

File: llvm/lib/Target/AArch64/SVEInstrFormats.td (hunk 1 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/SVEInstrFormats.td]: Done

**Review by CarolineConcatto — COMMENTED** [review]
Momchil, I believe the patch is fine, but I think you have two patches into one.
FCVT nad FDOTs. Maybe we should split.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 4]
```diff
@@ -10719,7 +10719,16 @@ Value *CodeGenFunction::EmitSVEDupX(Value *Scalar, llvm::Type *Ty) {
       cast<llvm::VectorType>(Ty)->getElementCount(), Scalar);
 }
 
-Value *CodeGenFunction::EmitSVEDupX(Value* Scalar) {
+Value *CodeGenFunction::EmitSVEDupX(Value *Scalar) {
+  if (auto *Ty = Scalar->getType(); Ty->isVectorTy()) {
```
[the remark this answers is not in the document]
Comment (momchil-velikov) [reply clang/lib/CodeGen/CGBuiltin.cpp:10723]: No, that would be inappropriate. It's an invariant, not an option.

[the remark this answers is not in the document]
Comment (momchil-velikov) [reply clang/lib/CodeGen/CGBuiltin.cpp:10723]: We can force termination in release mode too, but I don't think that's the practice in LLVM.

**Pull request merged**